tq/ssh.go

```diff
@@ -0,0 +1,386 @@
+					if err != nil {
+						return nil, fmt.Errorf("batch response: invalid expires-at: %s", entry)
+					}
+				}
+			}
+		}
+	}
+
+	for _, obj := range bRes.Objects {
+		obj.Missing = missing[obj.Oid]
+		for _, a := range obj.Actions {
+			a.createdAt = requestedAt
+		}
+	}
+
+	return bRes, nil
+}
+
+func (a *SSHBatchClient) MaxRetries() int {
+	return a.maxRetries
+}
+
+func (a *SSHBatchClient) SetMaxRetries(n int) {
+	a.maxRetries = n
+}
+
+type SSHAdapter struct {
+	*adapterBase
+	ctx      lfshttp.Context
+	transfer *ssh.SSHTransfer
+}
+
+// WorkerStarting is called when a worker goroutine starts to process jobs
+// Implementations can run some startup logic here & return some context if needed
+func (a *SSHAdapter) WorkerStarting(workerNum int) (interface{}, error) {
+	a.transfer.SetConnectionCountAtLeast(workerNum + 1)
+	return a.transfer.Connection(workerNum), nil
+}
+
+// WorkerEnding is called when a worker goroutine is shutting down
+// Implementations can clean up per-worker resources here, context is as returned from WorkerStarting
+func (a *SSHAdapter) WorkerEnding(workerNum int, ctx interface{}) {
+}
+
+func (a *SSHAdapter) tempDir() string {
+	// Shared with the basic download adapter.
+	d := filepath.Join(a.fs.LFSStorageDir, "incomplete")
+	if err := tools.MkdirAll(d, a.fs); err != nil {
+		return os.TempDir()
+	}
+	return d
+}
+
+// DoTransfer performs a single transfer within a worker. ctx is any context returned from WorkerStarting
+func (a *SSHAdapter) DoTransfer(ctx interface{}, t *Transfer, cb ProgressCallback, authOkFunc func()) error {
+	if authOkFunc != nil {
+		authOkFunc()
+	}
+	conn := ctx.(*ssh.PktlineConnection)
+	if a.adapterBase.direction == Upload {
+		return a.upload(t, conn, cb)
+	} else {
+		return a.download(t, conn, cb)
+	}
+}
+
+func (a *SSHAdapter) download(t *Transfer, conn *ssh.PktlineConnection, cb ProgressCallback) error {
+	rel, err := t.Rel("download")
+	if err != nil {
+		return err
+	}
+	if rel == nil {
+		return errors.Errorf("No download action for object: %s", t.Oid)
+	}
+	// Reserve a temporary filename. We need to make sure nobody operates on the file simultaneously with us.
+	f, err := tools.TempFile(a.tempDir(), t.Oid, a.fs)
+	if err != nil {
+		return err
+	}
+	tmpName := f.Name()
+	defer func() {
+		if f != nil {
+			f.Close()
+		}
+		os.Remove(tmpName)
+	}()
+
+	return a.doDownload(t, conn, f, cb)
+}
+
+// doDownload starts a download. f is expected to be an existing file open in RW mode
+func (a *SSHAdapter) doDownload(t *Transfer, conn *ssh.PktlineConnection, f *os.File, cb ProgressCallback) error {
+	args := a.argumentsForTransfer(t, "download")
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessage(fmt.Sprintf("get-object %s", t.Oid), args, false)
+	if err != nil {
+		return err
+	}
+	status, _, data, err := conn.ReadStatusWithData()
+	if err != nil {
+		return err
+	}
+	if status < 200 || status > 299 {
+		if data != nil {
+			io.Copy(ioutil.Discard, data)
```

Comment:
Good point.  I've parsed that data up to 1024 bytes and then appended it to the error message.

Reply:
This may be overly cautious on my part ... since we're dealing with an error condition here, what would happen if the server returned a huge blob of data in the place of the `error-message`?  Would there be any sense in using `io.CopyN(buffer, data, 1024)` first, to try to get no more than 1024 bytes, and then if it returns `err` as `nil` (i.e., there's more to read), using  `io.Copy(ioutil.Discard, data)` to slurp the rest?

Reply:
Sure, that seems fine.  There are probably other fun ways the server can DoS us, but that doesn't mean we need to add to them.